File: lab/scalability/bench.go
package main

import (
	"fmt"
	"os"
	"time"
)

const MAX_D = 100
const N = 1000000

func stage(in <-chan int) <-chan int {
	out := make(chan int)
	go func() {
		for v := range in {
			out <- v
		}
		close(out)
	}()
	return out
}

func main() {
	file, _ := os.Create("go_chain.txt")
	defer file.Close()

	for d := 1; d <= MAX_D; d++ {
		start := time.Now()

		for i := 0; i < N; i++ {
			var ch <-chan int

			base := make(chan int, 1)
			base <- 1
			close(base)

			ch = base

			for j := 0; j < d; j++ {
				ch = stage(ch)
			}

			for range ch {
			}
		}

		elapsed := time.Since(start)
		fmt.Printf("%d %f\n", d, elapsed.Seconds())
		fmt.Fprintf(file, "%d %f\n", d, elapsed.Seconds())
	}
}

File: lab/scalability/bench.py
from time import perf_counter

MAX_D = 100
N = 10**6

def base():
    yield 1

def make_chain(d):
    def gen():
        yield 1

    for _ in range(d):
        prev = gen
        def gen(prev=prev):
            yield from prev()
    return gen

depths = []
times = []

for d in range(1, MAX_D + 1):
    print(f"Testing depth {d}...")
    g = make_chain(d)()

    start = perf_counter()
    for _ in range(N):
        try:
            next(g)
        except StopIteration:
            g = make_chain(d)()
    t = perf_counter() - start

    depths.append(d)
    times.append(t)

with open("python_chain.txt", "w") as f:
    for d, t in zip(depths, times):
        f.write(f"{d} {t}\n")
